### PR TITLE
Fix iterator constructor and other failing tests

### PR DESCRIPTION
**Pull request by cjprybol**
Fixes https://github.com/BioJulia/Kmers.jl/issues/33 but depends on https://github.com/BioJulia/BioSequences.jl/pull/271 in order to pass all tests

```bash
$ julia runtests.jl 
Test Summary:          | Pass  Total
BioSequences Interface |    8      8
Test Summary:                | Pass  Total
Construction and Conversions |  260    260
Test Summary: | Pass  Total
Comparisons   |   86     86
Test Summary: | Pass  Total
Length        |   15     15
Test Summary:         | Pass  Total
Access and Iterations |   57     57
Test Summary: |  Pass  Total
Random        | 12448  12448
Test Summary: | Pass  Total
Find          |   33     33
Test Summary: | Pass  Total
Print         |    5      5
Test Summary: | Pass  Total
Show          |    5      5
Test Summary:   | Pass  Total
Transformations | 3848   3848
Test Summary: | Pass  Total
Mismatches    | 3200   3200
Test Summary: | Pass  Total
Matches       | 3200   3200
Test Summary:       | Pass  Total
De Bruijn Neighbors |    4      4
Test Summary: | Pass  Total
EveryKmer     |   14     14
Test Summary: | Pass  Total
SpacedKmers   |   14     14
Test Summary:      | Pass  Total
EveryCanonicalKmer |   18     18
Test Summary:        | Pass  Total
SpacedCanonicalKmers |   26     26
Test Summary: | Pass  Total
Translation   |   39     39
Test Summary: | Pass  Total
CodonSet      |  241    241
Test Summary:       | Pass  Total
Reverse translation |  148    148
```

### CODE REVIEW

File: src/kmer.jl [hunk 1]
```diff
@@ -156,7 +156,8 @@ function Kmer{A,K,N}(itr) where {A,K,N}
     # Construct the head.
     head = zero(UInt64)
     @inbounds for i in 1:n_head
-        sym = convert(eltype(Kmer{A,K,N}), itr[i])
```
Comment (cjprybol) [inline src/kmer.jl:159]: We're not able to index into an iterator, but we can iterate it at the index that we want

File: src/revtrans.jl [hunk 2]
```diff
@@ -1,5 +1,3 @@
-struct Unsafe end
```
Comment (cjprybol) [inline src/revtrans.jl:1]: This is a duplicate definition https://github.com/BioJulia/Kmers.jl/blob/5dd8f8ff122e538f200582561defc836a3272046/src/kmer.jl#L15

File: test/access.jl [hunk 3]
```diff
@@ -31,7 +31,8 @@
         @test iterate(DNAKmer("ACTG"), 1)  !== nothing
         @test iterate(DNAKmer("ACTG"), 4)  !== nothing
         @test iterate(DNAKmer("ACTG"), 5)  === nothing
-        @test_throws BoundsError iterate(DNAKmer("ACTG"), -1)
```
Comment (cjprybol) [inline test/access.jl:34]: Incorrect behavior per https://github.com/BioJulia/BioSequences.jl/pull/271

File: test/construction_and_conversion.jl [hunk 4]
```diff
@@ -8,6 +8,9 @@ global reps = 10
     @test DNAKmer(DNA_G, DNA_C, DNA_T) == Kmer("GCT")
     @test RNAKmer(RNA_G, RNA_U, RNA_C, RNA_U) == Kmer("GUCU")
     
+    # creation from iterator
+    @test Kmers.kmertype(Kmer{DNAAlphabet{2},31})((i for i in rand(ACGT, 31))) isa Kmers.kmertype(Kmer{DNAAlphabet{2},31})
```
Comment (cjprybol) [inline test/construction_and_conversion.jl:12]: Specifically tests the example in https://github.com/BioJulia/BioSequences.jl/issues/270

File: test/runtests.jl [hunk 5]
```diff
@@ -10,7 +10,7 @@ include("utils.jl")
 
 if GROUP == "BioSequences" || GROUP == "All"
     include("biosequences_interface.jl")
-    include("conversion.jl")
+    include("construction_and_conversion.jl")
```
Comment (cjprybol) [inline test/runtests.jl:13]: clarify contents of the file, which tests both constructors and conversions

File: src/kmer.jl [hunk 6]
```diff
@@ -167,7 +168,8 @@ function Kmer{A,K,N}(itr) where {A,K,N}
         Base.@_inline_meta
         body = zero(UInt64)
         @inbounds for i in 1:n_per_chunk
```
Comment (kescobo) [inline src/kmer.jl:170]: Not critical, but I think `for i in 1:x` is starting to be considered gauche due to things like offset arrays. Is there a way to do this with `firstindex` or similar (only if it doesn't require bending over backwards).

File: test/access.jl [hunk 7]
```diff
@@ -31,7 +31,8 @@
...
+        @test iterate(DNAKmer("ACTG"), -1) === nothing
```
Comment (kescobo) [inline test/access.jl:34]:
```suggestion
        @test isnothing(iterate(DNAKmer("ACTG"), -1))
```

File: test/access.jl [hunk 8]
```diff
@@ -67,7 +68,8 @@
         @test iterate(RNAKmer("ACUG"), 1)  !== nothing
```
Comment (kescobo) [inline test/access.jl:68]:
```suggestion
        @test !isnothing(iterate(RNAKmer("ACUG"), 1))
```
etc.

**Review by kescobo — CHANGES_REQUESTED** [review]
Rad, thanks for all the help with this package.! 

To be clear - it looks like the dependency you referenced at the beginning has been merged - is there a release of BioSequences that includes it yet? Once it is, this PR should include an update to the `[compat]` section of `Project.toml`.

**Comment by cjprybol** [comment]
> Rad, thanks for all the help with this package.!

Glad to help, thanks for the assistance with the PRs! 

> To be clear - it looks like the dependency you referenced at the beginning has been merged - is there a release of BioSequences that includes it yet? Once it is, this PR should include an update to the `[compat]` section of `Project.toml`.

I don't see a new release yet but happy to update the compat once one is

File: test/access.jl [hunk 9]
```diff
@@ -64,10 +65,11 @@
         @test iterate(RNAKmer("ACUG"), 4) == (RNA_G, 5)
         
 
-        @test iterate(RNAKmer("ACUG"), 1)  !== nothing
+        @test !isnothing(iterate(RNAKmer("ACUG"), 1))
         @test iterate(RNAKmer("ACUG"), 4)  !== nothing
         @test iterate(RNAKmer("ACUG"), 5)  === nothing
-        @test_throws BoundsError iterate(RNAKmer("ACUG"), -1)
+        @test iterate(RNAKmer("ACUG"), -1) === nothing
+        @test iterate(RNAKmer("ACUG"), 0) === nothing
```
Comment (kescobo) [inline test/access.jl:72]: Can we update all of these? My apologies - I didn't notice before that you were copying the style of tests that were already there. I think either revert my suggestion above, or change them all.

**Comment by kescobo** [comment]
@jakobnissen is there something holding back a release of BioSequences with https://github.com/BioJulia/BioSequences.jl/pull/271 ? I'm happy to do it, but wanted to double check.

**Comment by jakobnissen** [comment]
Nope, let's cut a release. Also nice to get the other changes since last release. They're all bugfixes, so it can be a patch release, 3.1.3.

**Comment by kescobo** [comment]
https://github.com/JuliaRegistries/General/pull/79281

@cjprybol Once that gets merged, I think bump the compat entry, which should also re-run the tests, and then I think it will be good to go

File: test/access.jl [hunk 10]
```diff
@@ -64,10 +65,11 @@
         @test iterate(RNAKmer("ACUG"), 4) == (RNA_G, 5)
         
 
-        @test iterate(RNAKmer("ACUG"), 1)  !== nothing
+        @test !isnothing(iterate(RNAKmer("ACUG"), 1))
```
Comment (kescobo) [inline test/access.jl:68]:
```suggestion
        @test iterate(RNAKmer("ACUG"), 1) !== nothing
```

don't want to just change the new one, my bad.

**Review by kescobo — APPROVED** [review]

**Comment by cjprybol** [comment]
Thanks for finishing this off @SabrinaJaye, and thanks for the helpful reviews @kescobo!

**Pull request merged**